algorithms/linfa-elasticnet/src/algorithm.rs

```diff
@@ -189,6 +319,92 @@ fn coordinate_descent<'a, F: Float>(
     (w, gap, n_steps)
 }
 
+fn block_coordinate_descent<'a, F: Float>(
+    x: ArrayView2<'a, F>,
+    y: ArrayView2<'a, F>,
+    tol: F,
+    max_steps: u32,
+    l1_ratio: F,
+    penalty: F,
+) -> (Array2<F>, F, u32) {
+    let n_samples = F::cast(x.shape()[0]);
+    let n_features = x.shape()[1];
+    let n_tasks = y.shape()[1];
+    // the parameters of the model
+    let mut w = Array2::<F>::zeros((n_features, n_tasks));
+    // the residuals: `Y - XW` (since W=0, this is just `Y` for now),
+    // the residuals are updated during the algorithm as the parameters change
+    let mut r = y.to_owned();
+    let mut n_steps = 0u32;
+    let norm_cols_x = x.map_axis(Axis(0), |col| col.dot(&col));
+    let mut gap = F::one() + tol;
+    let d_w_tol = tol;
+    let tol = tol * y.iter().map(|&y_ij| y_ij * y_ij).sum();
+    while n_steps < max_steps {
+        let mut w_max = F::zero();
+        let mut d_w_max = F::zero();
+        for j in 0..n_features {
+            if abs_diff_eq!(norm_cols_x[j], F::zero()) {
+                continue;
+            }
+            let mut old_w_j = w.slice_mut(s![j, ..]);
+            let x_j = x.slice(s![.., j]);
+            let norm_old_w_j = old_w_j.dot(&old_w_j).sqrt();
+            if abs_diff_ne!(norm_old_w_j, F::zero()) {
+                general_mat_mul(
+                    F::one(),
+                    &x_j.view().insert_axis(Axis(1)),
+                    &old_w_j.view().insert_axis(Axis(0)),
+                    F::one(),
+                    &mut r,
+                );
+                //for i in 0..x.shape()[0] {
```

Comment:
is this `general_mat_mul` previous version? If so then throw it away and add a comment before